### PR TITLE
feat: cart

### PR DESCRIPTION
**Pull request by Vishruta-Patil**
Features:
1. Cart UI
2. Fetch cart details of the user after login
3. Add to cart
4. Remove from Cart
5. Increment/ Decrement cart product quantity
6. Move product from Cart to Wishlist and vice versa

### CODE REVIEW

File: src/components/common/ProductCard/index.js [hunk 1]
```diff
@@ -26,16 +35,16 @@ const ProductCard = ({ item, clickHandler, from }) => {
...
-}
-        </button>      
+
+        <button style={{ background: "none", padding: 0 }} onClick={btnHandler}>
+          {
+            from ? (
+              <span class="material-icons wishlist-in-icon"> favorite </span>
+            ) : (
+              <span class={`material-icons ${activeClass}`}> favorite </span>
+            ) //state.wishListStatusClass
```
Comment (prathmeshjagtap) [inline src/components/common/ProductCard/index.js:45]: you can remove commented code

File: src/components/common/ProductCard/index.js [hunk 2]
```diff
@@ -45,11 +54,27 @@ const ProductCard = ({ item, clickHandler, from }) => {
...
+            Remove from Cart<span class="material-icons "> shopping_cart </span>
+          </button>
+        )}
       </div>
     </div>
   );
 };
 
 export default ProductCard;
+// onClick={({item}) => console.log({item})}
```
Comment (prathmeshjagtap) [inline src/components/common/ProductCard/index.js:80]: you can remove this line

File: src/components/home/categoryContainer.js [hunk 3]
```diff
@@ -13,4 +15,11 @@ const CategoryContainer = () => {
...
-export default CategoryContainer
\ No newline at end of file
+export default CategoryContainer
+
+
+//onClick={() =>
+// dispatch({
+//   type: FILTER_BY_CATEGORY,
+//   payload: categoryName.toLowerCase(),  // {categoryName}
+// })}
```
Comment (prathmeshjagtap) [inline src/components/home/categoryContainer.js:25]: remove this code if you are not going to use it in future

**Review by prathmeshjagtap — CHANGES_REQUESTED** [review]
In your Preview link Add to Cart Functionality doesn't seem to work, look into that

Update : Sorry my bad , it is working

**Review by prathmeshjagtap — APPROVED** [review]

**Review by abhijeetekad — APPROVED** [review]

File: src/components/common/HorizontalCard/index.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          <div class="price-tag flex">
+            <div class="price-amount">₹{item.product_price}</div>
+            <div class="duplicate-price-amount">₹3999</div>
+          </div>
+          <div class="price-offer-cart">{item.product_offer} off</div>
+          <div class="quantity-container flex">
+            <p>Quantity:</p>
+            <button
+              class="quantity-btn"
+              onClick={() => cartQuantityHandler(item, "increment", dispatch)}
```
Comment (rehanmohiuddin) [inline src/components/common/HorizontalCard/index.js:26]: increment,decrement can be be kept as contants like CART_INCREMENT

File: src/components/common/ProductCard/index.js [hunk 5]
```diff
@@ -2,21 +2,30 @@ import axios from "axios";
...
+  const [activeClass, setActiveClass] = useState("wishlist-icon");
+  const { state, dispatch } = useWishList();
 
   const btnHandler = () => {
     // dispatch({type: WISHLIST_STATUS_CLASS})
-    clickHandler({ item })
-    if(activeClass ==="wishlist-icon") {
-      setActiveClass("wishlist-active")
+    clickHandler({ item });
+    if (activeClass === "wishlist-icon") {
```
Comment (rehanmohiuddin) [inline src/components/common/ProductCard/index.js:14]: use ternary operator for single if else condition,if there are multiple then make a object with key as active class and value as classname

File: src/components/common/ProductCard/index.js [hunk 6]
```diff
@@ -2,21 +2,30 @@ import axios from "axios";
...
+      setActiveClass("wishlist-active");
     } else {
-      setActiveClass("wishlist-icon")
+      setActiveClass("wishlist-icon");
     }
-  }
+  };
+
+  const inCartHandler = (product) => {
+    const isCart = state.cartData.find((item) => item._id === product._id);
```
Comment (rehanmohiuddin) [inline src/components/common/ProductCard/index.js:22]: isCart?true:false you can do like this which will be much cleaner

File: src/components/common/ProductCard/index.js [hunk 7]
```diff
@@ -26,16 +35,16 @@ const ProductCard = ({ item, clickHandler, from }) => {
...
-          style={{ background: "none", padding: 0}}
-          onClick={btnHandler}
-        >
-          {from ?
-          <span class="material-icons wishlist-in-icon"> favorite </span>
-          :  <span class={`material-icons ${activeClass}`}> favorite </span> //state.wishListStatusClass
-}
-        </button>      
+
+        <button style={{ background: "none", padding: 0 }} onClick={btnHandler}>
```
Comment (rehanmohiuddin) [inline src/components/common/ProductCard/index.js:39]: **DO NOT DO INLINE STYLING PLEASE**

File: src/pages/Wishlist/index.js [hunk 8]
```diff
@@ -1,75 +1,50 @@
...
-          <div class="wishlist-unit">
-            {state.wishListData.map((data,index)=> (
-             <ProductCard item={data} key={index} clickHandler={deleteWishListHandler} from={true}/>
-            ))}
+          <div>
+            {state.wishListData.length === 0 ? (
+              <EmptyBag
+                name={"Wishlist"}
+                img={
+                  "https://res.cloudinary.com/debanftke/image/upload/v1648206283/wishlist_empty_pcbldx.jpg"
```
Comment (rehanmohiuddin) [inline src/pages/Wishlist/index.js:28]: for empty cart static images make a constant as EMPTY_BAG_IMG keep in seperate constants file and import it here

File: src/pages/auth/login.js [hunk 9]
```diff
@@ -1,38 +1,20 @@
...
-      dispatch({type: LOGIN_STATUS})
-      setTimeout(() => dispatch({ type: USER_LOADING }), 1000);
-    } catch (err) {
-      console.log("Error: ", err);
-    }
-  };
-
   const loginHandler = async () => {
-    await login(credentails.email, credentails.password);
+    await login(credentails.email, credentails.password, dispatch, navigate);
```
Comment (rehanmohiuddin) [inline src/pages/auth/login.js:13]: pass whole credentials as object credentials which makes params shorter to read and work

File: src/pages/auth/login.js [hunk 10]
```diff
@@ -1,38 +1,20 @@
...
   };
 
   const guestCredentailsHandler = async () => {
-    await login("adarshbalika@gmail.com", "adarshBalika123");
+    await login("adarshbalika@gmail.com", "adarshBalika123", dispatch, navigate);
```
Comment (rehanmohiuddin) [inline src/pages/auth/login.js:17]: make these as constants and import

File: src/pages/home/index.js [hunk 11]
```diff
@@ -44,14 +27,14 @@ export const Home = () => {
...
-
+ 
       <div class="top-products-container">
         <ContentTitle styles={"offer-header"}>
           MOST DAZZLING OFFERS
         </ContentTitle>
         <div class="product-container">
           {topProductsData.map(item => (
-            <ProductCard item={item} key={item._id} clickHandler={addToWishListHandler}/>
+            <ProductCard item={item} key={item._id} clickHandler={({item}) => addToWishListHandler({item})}/>
```
Comment (rehanmohiuddin) [inline src/pages/home/index.js:37]: for clickHandler function take into another function defined upwards don't make inline function definitions (a suggestion)

File: src/utils/handler.js [hunk 12]
```diff
@@ -1,16 +1,140 @@
...
+    dispatch({type: LOGIN_STATUS})
+    setTimeout(() => dispatch({ type: USER_LOADING }), 1000);
+  } catch (err) {
+    console.log("Error: ", err);
+  }
+};
+
+
+// Product List
+export const getProductList = async (dispatch) => {
```
Comment (rehanmohiuddin) [inline src/utils/handler.js:43]: every time for repeating url such as /api/ at top define in axios.defaults.baseUrl="/api/" then no need to repeat in every axios call (just a suggestion)

File: src/utils/handler.js [hunk 13]
```diff
@@ -1,16 +1,140 @@
...
+// Wishlist
+export const getWishlistItems = async (dispatch) => {
+  dispatch({ type: WISHLIST_LOADER });
+  try {
+    const response = await axios.get("/api/user/wishlist", config);
+    dispatch({ type: WISHLIST_LOADER });
+    dispatch({ type: WISHLIST_DATA, payload: response.data.wishlist });
+  } catch (err) {
+    dispatch({ type: WISHLIST_LOADER });
+    console.log("error: ", err);
```
Comment (rehanmohiuddin) [inline src/utils/handler.js:65]: remove console.log after debugging of code

**Pull request merged**